arch/armv7m/jmp.S

```diff
@@ -7,91 +7,116 @@
  * _setjmp, _longjmp, setjmp, sigsetjmp
  *
  * Copyright 2018, 2019 Phoenix Systems
- * Author: Kamil Amanowicz, Aleksander Kaminski
+ * Author: Kamil Amanowicz, Aleksander Kaminski, Jakub Sarzyński
  *
  * This file is part of Phoenix-RTOS.
  *
  * %LICENSE%
  */
 
+ /* setjmp env buffer description
+  * Offset	Description
+  * 0		Signal mask flag - if != 0 then next value stores a signal mask
+  * 4		Signal mask
+  * 8		SP register
+  * 12-44	r4-r11 registers
+  * 44		LR register
+  * 48-80	d8-d15 registers
+  * 80		fpscr register
+  */
+
 #define __ASSEMBLY__
 
 .thumb
 .syntax unified
 
-.globl _setjmp
-.type _setjmp, %function
-_setjmp:
-	/* set jump address and stack pointer */
-	ldr r1, =(1f + 1)
-	mov r2, sp
-	/* store everythig in jmpbuf */
-	stm r0, {r1-r2, r4-r11, lr}
+
+.globl sigsetjmp
+.type sigsetjmp, %function
+sigsetjmp:
+	/* Store signal mask if savesigs != 0 */
+	mov r2, #0
+	cmp r1, #0
+	beq 1f
+
+	/* Get signal mask */
+	sub sp, #4
```

Comment:
Why `sub sp, #4`?

Reply:
Stack must be aligned to 8 bytes before syscall

Reply:
Ok. I wanted to make sure.
Instead of:
```
sub sp, #4
...
add sp, #4 
```
can be done like this:
```
push {r0, r1, r3, lr}
...
pop {r0, r1, r3, lr}
```

Reply:
I am in favour of leaving here `sub/add` - these instructions say directly: "hey we need to change stack pointer, it must be aligned". Pushing/poping an unnecessary register may lead to confusion.

Wait for other reviews.

Reply:
push {r0, r1, r3, lr} version is better. We can't have stack unaligned at any time - interrupt can occur between sub and push - we'll enter irq with unaligned stack (some context is stored on user stack on interrupts).